tools/criticalstat_example.txt

```diff
@@ -0,0 +1,139 @@
+Demonstrations of criticalstat: Find long atomic critical sections in the kernel.
+
+criticalstat traces and reports occurences of atomic critical sections in the
+kernel with useful stacktraces showing the origin of them. Such critical
+critical sections frequently occur due to use of spinlocks, or if interrupts or
+preemption were explicity disabled by a driver. IRQ routines in Linux are also
+executed with interrupts disabled. There are many reasons. Such critical
+sections are a source of long latency/responsive issues for real-time systems.
+
+This works by probing the preempt/irq and cpuidle tracepoints in the kernel.
+Since this uses BPF, only the root user can use this tool. Further, the kernel
+has to be built with certain CONFIG options enabled inorder for it to work:
+CONFIG_PREEMPTIRQ_EVENTS
+CONFIG_DEBUG_PREEMPT
+Additionally, the following options should be turned off on older kernels:
+CONFIG_PROVE_LOCKING
+CONFIG_LOCKDEP
+
+USAGE:
+# ./criticalstat -h
+usage: criticalstat [-h] [-p] [-i] [-d DURATION]
+
+Trace long critical sections
+
+optional arguments:
+  -h, --help            Show this help message and exit
+  -p, --preemptoff      Find long sections where preemption was off
+  -i, --irqoff          Find long sections where IRQ was off
+  -d DURATION, --duration DURATION
+                        Duration in uS (microseconds) below which we filter
+
+examples:
+    ./criticalstat          	# run with default options: irq off for more than 100 uS
+    ./criticalstat -p       	# find sections with preemption disabled for more than 100 uS
+    ./criticalstat -d 500   	# find sections with IRQs disabled for more than 500 uS
+    ./criticalstat -p -d 500	# find sections with preemption disabled for more than 500 uS
+
+The tool runs continuously until interrupted by Ctrl-C
+
+
+By default, criticalstat finds IRQ disable sections for > 100us.
+
+# ./criticalstat
+Finding critical section with IRQ disabled for > 100us
+===================================
+TASK: kworker/u16:5 (pid  5903 tid  5903) Total Time: 194.427  us
+
+Section start: __schedule -> schedule
+Section end:   _raw_spin_unlock_irq -> finish_task_switch
+  trace_hardirqs_on+0xdc
+  trace_hardirqs_on+0xdc
```

Comment:
Why we have two `trace_hardirqs_on+0xdc`? The same for some examples below.

Reply:
Yeah, not really sure why. This is something to do with how stack unwinder's internals. It also appear the first couple of frames are ignored by other stack dumping methods (such as ftrace event's stacktrace trigger)

For example, in comparison:
Here's ftrace stacktrace trigger on the preempt enable event:
```
Section start: perf_event_read -> perf_event_read
Section end:   perf_event_read -> perf_event_read
  trace_preempt_on+0x98
  trace_preempt_on+0x98
  preempt_count_sub+0xa4
  perf_event_read+0x2cc
  perf_event_read_value+0x44
  get_cnt+0xb4
  devfreq_memlat_get_freq+0x44
  update_devfreq+0x64
  devfreq_monitor+0x30
  process_one_work+0x218
  worker_thread+0x34c
  kthread+0x118
  ret_from_fork+0x10
```
In comparison, here is the one from criticalstat (this tool):
```
Section start: perf_event_read -> perf_event_read
Section end:   perf_event_read -> perf_event_read
  trace_preempt_on+0x98
  trace_preempt_on+0x98
  preempt_count_sub+0xa4
  perf_event_read+0x2cc
  perf_event_read_value+0x44
  get_cnt+0xb4
  devfreq_memlat_get_freq+0x44
  update_devfreq+0x64
  devfreq_monitor+0x30
  process_one_work+0x218
  worker_thread+0x34c
  kthread+0x118
  ret_from_fork+0x10
```
The stack itself is correct though since it has all the symbols so I am Ok with it for now but we should fix this issue later on if possible.